### PR TITLE
fix: set max option not work when send from qrcode

### LINKED ISSUE #11460: 'Set max' option is not working if "Send transaction" screen was opened using universal scanner
# Bug Report

## Problem
When scanning QR code with wallet address using universal QR code scanner or QR code scanner in wallet, "Set max" option does nothing.
BAACAgQAAxkBAAEtg2Nft74ea1UWPnrqdV9KPzcsFfC8QQACyggAApKSwVE6GTjljgizgh4E_1605877392 96

#### Expected behavior
can set max

#### Actual behavior
nothing happens

### Reproduction

[comment]: # (Describe how we can replicate the bug step by step.)

- Open Status
- Using "Universal QR scanner" scan valid wallet address
- Try to "Set max"

### Additional Information

[comment]: # (Please do your best to fill this out.)
- Status version: nightly 20/11/20 
- Operating System: Android, iOS

### PR DESCRIPTION
**Pull request by yqrashawn**
### Summary

resolve #11460

The bug is caused by lack of `[:wallet/prepare-transaction :gasPrice]` in db
code at: https://github.com/status-im/status-mobile/blob/7a97d88983cce567b781619ad379cf96227d1ecf/src/status_im/wallet/core.cljs#L390-L395

which is caused by these two functions https://github.com/status-im/status-mobile/blob/7a97d88983cce567b781619ad379cf96227d1ecf/src/status_im/wallet/choose_recipient/core.cljs#L54-L104

`fill-prepare-transaction-details` is used to fill `:wallet/prepare-transaction` based on `details` parse from URI (parsed from qr code here)
which won't fill `:gasPrice` if it's not presented in qr code

the solution here is to return `:signing/update-gas-price` fx if there's no `[:wallet/prepare-transaction :gasPrice]` in the result of `fill-prepare-transaction-details`

#### Platforms

- Android
- iOS

#### Areas that maybe impacted

logic related to [eip-681](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-681.md) or this function https://github.com/status-im/status-mobile/blob/7a97d88983cce567b781619ad379cf96227d1ecf/src/status_im/wallet/choose_recipient/core.cljs#L113

### Steps to test

- Open Status
- Using "Universal QR scanner" scan valid wallet address
- Try to "Set max"

status: ready

### CODE REVIEW

**Comment by yqrashawn** [comment]
before:

https://user-images.githubusercontent.com/15090582/181487946-674b5884-9570-4495-b964-6b22258b9140.mp4

after:

https://user-images.githubusercontent.com/15090582/181487968-24ac2e31-e45f-458c-ac6c-01b4a12b68ce.mp4

**Comment by yqrashawn** [comment]
And although there are no keys like `maxPriorityFeePerGas` in eip-681
we might need to get them handled

File: shadow-cljs.edn [hunk 1]
```diff
@@ -7,7 +7,7 @@
                 [binaryage/oops "0.7.0"]
                 [com.andrewmcveigh/cljs-time "0.5.2"]
                 [status-im/timbre "4.10.0-2-status"]
-                [com.taoensso/encore "2.105.0"]
+                [com.taoensso/encore "3.24.0"]
```
Comment (flexsurfer) [inline shadow-cljs.edn]: could you pls elaborate on why this change is needed for this PR ?

Comment (yqrashawn) [reply shadow-cljs.edn]: Oh, that's not necessary.
I saw it's not used else where. So I thought I'll just upgrade it to the latest version, which have bug fixes and improvements, and it won't break anything.

Do I need to change it back?

Comment (flexsurfer) [reply shadow-cljs.edn]: yeah its better to rever it, thanks

File: src/status_im/wallet/choose_recipient/core.cljs [hunk 2]
```diff
@@ -75,22 +76,29 @@
...
-                       :to address :to-name (find-address-name db address))}
-          (let [current-chain-id (get-in networks [current-network :config :NetworkId])]
-            (merge {:db (fill-prepare-transaction-details db details all-tokens)
-                    :dispatch [:open-modal :prepare-send-transaction]}
-                   (when (and chain-id (not= current-chain-id chain-id))
-                     {:ui/show-error (i18n/label :t/wallet-invalid-chain-id
-                                                 {:data uri :chain current-chain-id})})))))
-      {:ui/show-error (i18n/label :t/wallet-invalid-address {:data uri})})))
+  (enc/cond
+    :let [{:keys [address] :as details}
```
Comment (flexsurfer) [inline src/status_im/wallet/choose_recipient/core.cljs]: could you elaborate pls on why `enc/cond`  is used, what is the point? thanks

Comment (yqrashawn) [reply src/status_im/wallet/choose_recipient/core.cljs]: Less nested and more readable code

Comment (yqrashawn) [reply src/status_im/wallet/choose_recipient/core.cljs]: Hi, I'm ok to change this back and use `if` instead. What do you think?

File: shadow-cljs.edn (hunk 1 above)
Comment (yqrashawn) [reply shadow-cljs.edn]: The version is reverted

File: src/status_im/wallet/choose_recipient/core.cljs (hunk 2 above)
Comment (flexsurfer) [reply src/status_im/wallet/choose_recipient/core.cljs]: it's better to use `if`, because it's more readable if are not familiar with `enc/cond`

Comment (yqrashawn) [reply src/status_im/wallet/choose_recipient/core.cljs]: I just removed usage of `enc/cond` in the latest commit

File: src/status_im/wallet/choose_recipient/core.cljs [hunk 3]
```diff
@@ -84,9 +84,17 @@
         (if (:wallet/prepare-transaction db)
           {:db (update db :wallet/prepare-transaction assoc
                        :to address :to-name (find-address-name db address))}
-          (let [current-chain-id (get-in networks [current-network :config :NetworkId])]
-            (merge {:db (fill-prepare-transaction-details db details all-tokens)
+          (let [current-chain-id                     (get-in networks [current-network :config :NetworkId])
+                new-db-with-transaction-from-details (fill-prepare-transaction-details db details all-tokens)]
+            (merge {:db       new-db-with-transaction-from-details
                     :dispatch [:open-modal :prepare-send-transaction]}
+                   (when-not (get-in new-db-with-transaction-from-details [:wallet/prepare-transaction :gasPrice])
```
Comment (flexsurfer) [inline src/status_im/wallet/choose_recipient/core.cljs]: i guess you could just check it in details ?

Comment (flexsurfer) [reply src/status_im/wallet/choose_recipient/core.cljs]:
```
[{:keys [address gasPrice] :as details}
        (eip681/extract-request-details data all-tokens)]
        ```

Comment (yqrashawn) [reply src/status_im/wallet/choose_recipient/core.cljs]: fixed

**Review by flexsurfer — APPROVED** [review]
thank you

**Comment by pavloburykh** [comment]
@yqrashawn thanx for the fix. Tested and ready for merge.
